xds/src/main/java/io/grpc/xds/EdsLoadBalancer2.java

```diff
@@ -0,0 +1,421 @@
+      return new ChildLbState(helper);
+    }
+
+    private final class ChildLbState extends LoadBalancer implements EndpointWatcher {
+      @Nullable
+      private final LoadStatsStore loadStatsStore;
+      private final DropHandlingLbHelper lbHelper;
+      private List<EquivalentAddressGroup> endpointAddresses = Collections.emptyList();
+      private Map<Integer, Map<Locality, Integer>> prioritizedLocalityWeights
+          = Collections.emptyMap();
+      private PolicySelection localityPickingPolicy;
+      private PolicySelection endpointPickingPolicy;
+      @Nullable
+      private LoadBalancer lb;
+
+      private ChildLbState(Helper helper) {
+        if (lrsServerName != null) {
+          loadStatsStore = xdsClient.addClientStats(cluster, edsServiceName);
+          xdsClient.reportClientStats();
+        } else {
+          loadStatsStore = null;
+        }
+        lbHelper = new DropHandlingLbHelper(helper);
+        logger.log(
+            XdsLogLevel.INFO,
+            "Start endpoint watcher on {0} with xDS client {1}", resourceName, xdsClient);
+        xdsClient.watchEndpointData(resourceName, this);
+      }
+
+      @Override
+      public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+        EdsConfig config = (EdsConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+        if (lb != null) {
+          if (!config.localityPickingPolicy.equals(localityPickingPolicy)
+              || !config.endpointPickingPolicy.equals(endpointPickingPolicy)) {
+            PriorityLbConfig childConfig =
+                generatePriorityLbConfig(cluster, edsServiceName, lrsServerName,
+                    config.localityPickingPolicy, config.endpointPickingPolicy, lbRegistry,
+                    prioritizedLocalityWeights);
+            // TODO(chengyuanzhang): to be deleted after migrating to use XdsClient API.
+            Attributes attributes;
+            if (lrsServerName != null) {
+              attributes =
+                  resolvedAddresses.getAttributes().toBuilder()
+                      .set(XdsAttributes.ATTR_CLUSTER_SERVICE_LOAD_STATS_STORE, loadStatsStore)
+                      .build();
+            } else {
+              attributes = resolvedAddresses.getAttributes();
+            }
+            lb.handleResolvedAddresses(
+                resolvedAddresses.toBuilder()
+                    .setAddresses(endpointAddresses)
+                    .setAttributes(attributes)
+                    .setLoadBalancingPolicyConfig(childConfig)
+                    .build());
+          }
+        }
+        localityPickingPolicy = config.localityPickingPolicy;
+        endpointPickingPolicy = config.endpointPickingPolicy;
+      }
+
+      @Override
+      public void handleNameResolutionError(Status error) {
+        if (lb != null) {
+          lb.handleNameResolutionError(error);
+        } else {
+          lbHelper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
```

Comment:
It shouldn't be this `DropHandlingLbHelper  lbHelper` either, that's for `lb` with dropping logic, it's too downstream. The right one should be the helper in `ChildLbState(Helper helper)`.

Reply:
Yeah, I know. It doesn't matter though. `DropHandlingLbHelper` is just a the helper received in `ChildLbState(Helper helper)` with a thin layer wrapped to record drops, which doesn't affect its error propagation functionality.

Reply:
From the API's perspective, your thought is legit. What about changing to `lbHelper.delegate().updateBalancingState(...)`? or `lbHelper.helper.updateBalancingState(...)`